test/parallel/test-common-can-not-mutate-object-deep.mjs

```diff
@@ -0,0 +1,199 @@
+}
+
+function setOnFoo(root) {
+  root.foo.xyzzy = 'gwak';
+}
+
+function setQux(root) {
+  root.qux = 'gwak';
+}
+
+function setQuux(root) {
+  root.quux.push('gwak');
+}
+
+function setQuuxItem(root) {
+  root.quux[0] = 'gwak';
+}
+
+function setQuuxProperty(root) {
+  root.quux[1].corge = 'gwak';
+}
+
+function setPrototypeOfRoot(root) {
+  Object.setPrototypeOf(root, Array);
+}
+
+function setPrototypeOfFoo(root) {
+  Object.setPrototypeOf(root.foo, Array);
+}
+
+function setPrototypeOfQuux(root) {
+  Object.setPrototypeOf(root.quux, Array);
+}
+
+
+{
+  assert.ok(testOriginal(obj()));
+
+  assert.throws(
+    () => definePropertyOnRoot(obj()),
+    { code: 'ERR_ASSERTION' }
+  );
+  assert.throws(
+    () => definePropertyOnFoo(obj()),
+    { code: 'ERR_ASSERTION' }
+  );
+  assert.throws(
+    () => deletePropertyOnRoot(obj()),
+    { code: 'ERR_ASSERTION' }
+  );
+  assert.throws(
+    () => deletePropertyOnFoo(obj()),
+    { code: 'ERR_ASSERTION' }
+  );
+  assert.throws(
+    () => preventExtensionsOnRoot(obj()),
+    { code: 'ERR_ASSERTION' }
+  );
+  assert.throws(
+    () => preventExtensionsOnFoo(obj()),
+    { code: 'ERR_ASSERTION' }
+  );
+  assert.throws(
+    () => preventExtensionsOnRootViaSeal(obj()),
+    { code: 'ERR_ASSERTION' }
+  );
+  assert.throws(
+    () => preventExtensionsOnFooViaSeal(obj()),
+    { code: 'ERR_ASSERTION' }
+  );
+  assert.throws(
+    () => preventExtensionsOnRootViaFreeze(obj()),
+    { code: 'ERR_ASSERTION' }
+  );
+  assert.throws(
+    () => preventExtensionsOnFooViaFreeze(obj()),
+    { code: 'ERR_ASSERTION' }
+  );
+  assert.throws(
+    () => setOnRoot(obj()),
+    { code: 'ERR_ASSERTION' }
+  );
+  assert.throws(
+    () => setOnFoo(obj()),
+    { code: 'ERR_ASSERTION' }
+  );
+  assert.throws(
+    () => setQux(obj()),
+    { code: 'ERR_ASSERTION' }
+  );
+  assert.throws(
+    () => setQuux(obj()),
+    { code: 'ERR_ASSERTION' }
+  );
+  assert.throws(
+    () => setQuux(obj()),
+    { code: 'ERR_ASSERTION' }
+  );
+  assert.throws(
+    () => setQuuxItem(obj()),
+    { code: 'ERR_ASSERTION' }
+  );
+  assert.throws(
+    () => setQuuxProperty(obj()),
+    { code: 'ERR_ASSERTION' }
+  );
+  assert.throws(
+    () => setPrototypeOfRoot(obj()),
+    { code: 'ERR_ASSERTION' }
+  );
+  assert.throws(
+    () => setPrototypeOfFoo(obj()),
+    { code: 'ERR_ASSERTION' }
+  );
+  assert.throws(
+    () => setPrototypeOfQuux(obj()),
+    { code: 'ERR_ASSERTION' }
+  );
+
+  // Test that no mutation happened
+  assert.ok(testOriginal(obj()));
+}
+
+// Test various supported types, directly and nested
+[
+  undefined, null, false, true, 42, 42n, Symbol('42'), NaN, Infinity, {}, [],
+  () => {}, async () => {}, Promise.resolve(), Math, Object.create(null),
+].forEach((target) => {
+  assert.deepStrictEqual(canNotMutateObjectDeep(target), target);
+  assert.deepStrictEqual(canNotMutateObjectDeep({ target }), { target });
+  assert.deepStrictEqual(canNotMutateObjectDeep([ target ]), [ target ]);
+});
```

Comment:
I was thinking of tests covering specifically the function case, e.g.:

```suggestion
});

{
  function target() {}
  target.property = {};

  // We want the following to keep working:
  assert.strictEqual(canNotMutateObjectDeep(target), target);
  util.callbackify(target); // callbackify/promisify should not throw.

  // We don't want the following to keep working, it conflicts with the above:
  target.property.key = 'value';
}

Reply:
Added function-specific tests via https://github.com/nodejs/node/pull/43196/commits/be6b1f6d84ac590247e910fc314be45b73f04494.
I think, tests for what we don't want to keep working are appropriate only when it's an unintended&undocumented feature that someone might rely on; but this one is no more than known limitation.
Instead, it tests that this limitation is not "fixed" in a wrong way such as freezing or replacing with a stub.